tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java

```diff
@@ -780,6 +759,147 @@ private ApplicationRpcServer setupRPCService(String hostname) {
     return rpcServer;
   }
 
+  private class TaskScheduler {
+    private Map<TensorFlowContainerRequest, Map<String, Integer>> taskDependencyMap = new HashMap<>();
+    boolean dependencyCheckPassed = true;
+
+    private void scheduleTasks() {
+      session.setResources(yarnConf, hdfsConf, localResources, containerEnv, hdfsClasspath);
+      final List<TensorFlowContainerRequest> requests = session.getContainersRequests();
+
+      if (!isDAG(requests)) {
+        LOG.error("TonY execution graph does not form a DAG, exiting.");
+        session.setFinalStatus(FinalApplicationStatus.FAILED, "TonY task scheduling failed..");
+        dependencyCheckPassed = false;
+        return;
+      }
+
+      // build task dependency graph
+      buildTaskDependency(requests);
+
+      // start/schedule jobs that have no dependency requirements
+      for (TensorFlowContainerRequest request : requests) {
+        if (checkDependencySatisfied(request)) {
+          scheduleJob(request);
+        }
+      }
+    }
+
+    private void buildTaskDependency(List<TensorFlowContainerRequest> requests) {
+      for (TensorFlowContainerRequest request : requests) {
+        for (String dependsOn : request.getDependsOn()) {
+          if (!dependsOn.isEmpty()) {
+            taskDependencyMap.putIfAbsent(request, new HashMap<>());
```

Comment:
We only construct taskDependencyMap here, so I think you can take this bit out, and in the next line: `Map<String, Integer> dependenciesForTask = new HashMap<>();`

Reply:
A task may depend on multiple other tasks, therefore some dependency might already exist in the taskDependencyMap.